modules/oidcsso/src/main/java/org/wso2/carbon/tomcat/oidcsso/extension/agent/OIDCAgent.java

```diff
@@ -0,0 +1,367 @@
+import com.nimbusds.openid.connect.sdk.AuthenticationErrorResponse;
+import com.nimbusds.openid.connect.sdk.AuthenticationResponseParser;
+import com.nimbusds.openid.connect.sdk.AuthenticationSuccessResponse;
+import com.nimbusds.openid.connect.sdk.LogoutRequest;
+import com.nimbusds.openid.connect.sdk.OIDCTokenResponse;
+import com.nimbusds.openid.connect.sdk.UserInfoErrorResponse;
+import com.nimbusds.openid.connect.sdk.UserInfoRequest;
+import com.nimbusds.openid.connect.sdk.UserInfoSuccessResponse;
+import com.nimbusds.openid.connect.sdk.claims.IDTokenClaimsSet;
+import com.nimbusds.openid.connect.sdk.validators.IDTokenValidator;
+
+import net.minidev.json.JSONObject;
+
+import org.wso2.carbon.tomcat.oidcsso.extension.Constants;
+import org.wso2.carbon.tomcat.oidcsso.extension.bean.AuthenticationResponse;
+import org.wso2.carbon.tomcat.oidcsso.extension.bean.RequestParameters;
+import org.wso2.carbon.tomcat.oidcsso.extension.bean.TokenResponse;
+import org.wso2.carbon.tomcat.oidcsso.extension.bean.UserInfoResponse;
+import org.wso2.carbon.tomcat.oidcsso.extension.oidc.StateStore;
+import org.wso2.carbon.tomcat.oidcsso.extension.utils.OIDCConfiguration;
+import org.wso2.carbon.tomcat.oidcsso.extension.utils.exception.AuthenticationRequestException;
+import org.wso2.carbon.tomcat.oidcsso.extension.utils.exception.AuthenticationResponseException;
+import org.wso2.carbon.tomcat.oidcsso.extension.utils.exception.LogoutException;
+import org.wso2.carbon.tomcat.oidcsso.extension.utils.exception.TokenException;
+import org.wso2.carbon.tomcat.oidcsso.extension.utils.exception.UserInfoException;
+
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.cert.Certificate;
+import java.security.cert.CertificateException;
+import java.security.interfaces.RSAPublicKey;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import javax.servlet.http.HttpServletRequest;
+
+/**
+ * This java class defines the API methods to complete OpenID Connect based SSO and SLO
+ */
+public class OIDCAgent {
+
+    /**
+     * this method builds the OpenID Connect Authentication Request
+     *
+     * @param oidcConfiguration web application specific OIDC configuration
+     * @param requestParameters parameters which defined in the request to authenticate from user
+     * @param stateStore        storage of state values
+     * @return Authentication Request String
+     * @throws AuthenticationRequestException if any error occurs during building the request
+     */
+    public String buildAuthenticationRequest(OIDCConfiguration oidcConfiguration, RequestParameters requestParameters,
+                                             StateStore stateStore) throws AuthenticationRequestException {
```

Comment:
It might be good practice to keep an empty line after the preceding line had to be split into two lines. This improves code readability.